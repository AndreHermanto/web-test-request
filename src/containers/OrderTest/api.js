/**
 * Returns the categorise of test panels including whole genome analysis.
 * @return {Object[]} json Currently returns an array of test panel object - including id and label
 */
export function getPanelsData() {
  return fetch(`${process.env.REACT_APP_BASE_URL}/panel-types`)
    .then((response) => response.json())
    .then((json) => json.data);
}
