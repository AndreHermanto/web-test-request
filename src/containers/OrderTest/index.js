import React, { Component } from 'react';
import {
  FormGroup,
  FormControl,
  ControlLabel,
  Label,
  Well
} from 'react-bootstrap';
import { 
  getTestList,
  getTest
} from './api';
import {
  setTestList,
  setTestType
} from './reducer'
import { 
  PageHeading,
  FormButton
} from './../../components/SharedStyle';
import styled from 'styled-components';

const Gene = styled(Label)`
  margin-right: 4px;
  display: inline-block !important;
  font-weight: 300 !important;
`;


/**
 * OrderTest - UI for ordering type of tests, selecting disorder and related genes for testing.
 */
class OrderTest extends Component {
  constructor(props) {
    super(props);
    
    this.handleTestSelect = this.handleTestSelect.bind(this);
    this.handleConfirm = this.handleConfirm.bind(this);

    this.state = {
      testList: [],
      form: props.route.data || {}
    };
  }
  
  componentWillMount() {
    return getTestList()
      .then((tests) => {
        this.setState(setTestList(this.state, tests));
      });
  }
  
  handleTestSelect(event) {
    if(event.target.value === 'whole') {
      var test = {
        "id": "whole",
        "label": "Whole Genome Analysis",
        "description": "Analysis on the entire genome."
      };

      this.setState(setTestType(this.state, test));
      return;
    }
    
    return getTest(event.target.value)
      .then((test) => { 
        this.setState(setTestType(this.state, test));
      });
  }
  
  handleConfirm() {
    this.props.route.onChange(this);
    this.props.router.push('/step2');
  }

  render() {
    return (
      <div>
        <PageHeading>Step 1: Order a Test</PageHeading>
        <FormGroup>
          <ControlLabel>Select a Disease Panel / Whole Genome Analysis</ControlLabel>
          <FormControl 
            componentClass="select" 
            name="test"
            value={this.state.form.test ? this.state.form.test.id : ''}
            onChange={this.handleTestSelect}
          >
            <option key="0" value="" disabled>Select a panel.</option> 
      
            {this.state.testList && (
              this.state.testList.map(
                (test) => {
                  return <option key={test.id} value={test.id}>{test.label}</option>
                }
              )
            )}
  
            <option key="whole" value="whole">Whole Genome Analysis</option>
          </FormControl>
        </FormGroup>
          
        {this.state.form.test && (
          <div style={{ marginBottom: 24 }}>
            {this.state.form.test.description}
          </div>
        )}
          

        {this.state.form.genes && (
          <div> 
            <ControlLabel>Available Genes:</ControlLabel>
            <Well>
            {this.state.form.genes.map((gene, $index) => {
              return <Gene key={$index}>{gene}</Gene> 
            })}
            </Well>
          </div>
        )}


        <FormButton 
          bsStyle="success" 
          type="submit" 
          onClick={this.handleConfirm}
        >
          Confirm
        </FormButton> 
      </div>
    );
  }
}

export default OrderTest;