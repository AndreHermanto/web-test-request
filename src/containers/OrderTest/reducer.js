import validator from './../../components/validator';

export function initData(prefilled) {
  var state = {
    form: {
      test: {}
    },
    testList: [],
    validationRule: {
      test: 'requireSelect',
    },
    validated: false,
    formId: 'orderTestModule'
  };
  
  if(prefilled && Object.keys(prefilled).length !== 0) state.form = prefilled;
  
  // This validates the data in the initial state.
  state.validation = {};
  for (var field in state.validationRule) {
    if(field) {
      state.validation[field] = validator(state.form[field], state.validationRule[field]);
    }
  }
  
  return Object.assign({}, state);
}


/**
 * This overrides the test list.
 * @param {Object} state Targeted state to be changed.
 * @param {string[]} list List of tests.
 */
export function setTestList(state, list) {
  return Object.assign({}, state, { 
    testList: list,
    formId: 'orderTestModule'
  });
}

/**
 * This sets the test type.
 * @param {Object} state Targeted state to be changed.
 * @param {string} name Name of the child in the form state.
 * @param {string} value Value for updating.
 */
export function setTestType(state, value) {
  var formStateChild = Object.assign({}, state.form, {
    "test": value
  });
  
  var validationStateChild = Object.assign({}, state.validation, {
    test: validator(value, state.validationRule['test'])
  });
  
  return Object.assign({}, state, {
    form: formStateChild,
    validation: validationStateChild
  });
}


/**
 * Set "validated" state to true - identifying the confirm button is clicked and validation processed.
 * @param {Object} state Targeted state to be changed.
 */
export function validatedToTrue(state) {
  return Object.assign({}, state, {
    validated: true
  });
}