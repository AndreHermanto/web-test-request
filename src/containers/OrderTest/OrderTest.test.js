import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TestUtils from 'react-addons-test-utils';
import FetchMock from 'fetch-mock';
import OrderTest from './index';

describe('OrderTest: index', () => {
  var testList = { data:[
    {id: 1, label: 'Test1'},
    {id: 2, label: 'Test2'},
    {id: 3, label: 'Test3'}
  ] };
  
  var props = {
    route: {
      onChange: jest.fn(),
      data: {}
    },
    router:['/step1']
  };
  
  var tests = { data: {id: 2, label: 'Test2', description: 'Test2', genes: ['ABC']} };
  
  test('renders without crashing - initial state with empty geneList', () => {
    const page = renderer.create(React.createElement(OrderTest, props)).toJSON();
    expect(page).toMatchSnapshot();
  });
  
  test('companyWillMount runs the function getting the test list', async () => {
    const page = TestUtils.renderIntoDocument(React.createElement(OrderTest, props));
    FetchMock.get('*', testList);
    await page.componentWillMount();
    expect(page.state.testList[1].id).toEqual(2);
    FetchMock.restore();                           
  });
  
  test('handleTestSelect works - select a test panel', () => {
    const page = TestUtils.renderIntoDocument(React.createElement(OrderTest, props));
    page.state.testList = testList.data;
    const selection = { "target": { "value": testList.data[1].id }};
    page.handleTestSelect(selection);
    expect(page.state.form.test.id).toEqual(2);  
  });
  
  test('handleConfirm works without selection', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(OrderTest, props));                                         
    page.handleConfirm();
    expect(page.props.router.pop()).toEqual('/step1')
  });
  
  test('handleConfirm works after selection', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(OrderTest, props));
    page.state.testList = testList.data;
    const selection = { "target": { "value": testList.data[1].id }};
    page.handleTestSelect(selection);
    page.handleConfirm();
    expect(page.props.router.pop()).toEqual('/step2')
  });
  
  test('handleConfirm works after selection in edit mode', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(OrderTest, props));
    page.state.testList = testList.data;
    page.props.route.isEdited = true;
    const selection = { "target": { "value": testList.data[1].id }};
    page.handleTestSelect(selection);
    page.handleConfirm();
    expect(page.props.router.pop()).toEqual('/summary')
  });
  
  test('displayGenes shows a list of genes', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(OrderTest, props));
    page.state.form.genes = ['AAAA', 'BBBB', 'CCCC']
    expect(page.displayGenes().length).toEqual(3)
  });
  
});
