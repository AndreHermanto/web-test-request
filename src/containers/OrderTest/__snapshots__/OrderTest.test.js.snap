exports[`OrderTest: index renders without crashing - initial state with empty geneList 1`] = `
<div>
  <h3
    className="jhCkYx">
    Step 1: Order a Test
  </h3>
  <div
    className="form-group"
    onChange={[Function]}>
    <label
      className="control-label"
      htmlFor={undefined}>
      Select a Disease Panel / Whole Genome Analysis
      <span
        className="jCNiHk">
        Required
      </span>
    </label>
    <br />
    <span
      className="SfhHK help-block" />
  </div>
  <button
    className="iMtXso btn btn-success"
    disabled={false}
    onClick={[Function]}
    type="submit">
    Confirm
  </button>
</div>
`;
